catalog/rocks/pgrefs.go

```diff
@@ -0,0 +1,222 @@
+		}, nil
+	}, db.ReadOnly(), db.WithContext(ctx))
+	if errors.Is(err, db.ErrNotFound) {
+		return nil, ErrNotFound
+	}
+	if err != nil {
+		return nil, err
+	}
+	return branch.(*Branch), nil
+}
+
+func (m *PGRefManager) SetBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID, branch Branch) error {
+	_, err := m.db.Transact(func(tx db.Tx) (interface{}, error) {
+		_, err := tx.Exec(`
+			INSERT INTO kv_branches (repository_id, id, staging_token, commit_id)
+			VALUES ($1, $2, $3, $4)
+				ON CONFLICT (repository_id, id)
+				DO UPDATE SET staging_token = $3, commit_id = $4`,
+			repositoryID, branchID, branch.stagingToken, branch.CommitID)
+		return nil, err
+	}, db.WithContext(ctx))
+	return err
+}
+
+func (m *PGRefManager) DeleteBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID) error {
+	_, err := m.db.Transact(func(tx db.Tx) (interface{}, error) {
+		_, err := tx.Exec(
+			`DELETE FROM kv_branches WHERE repository_id = $1 AND id = $2`,
+			repositoryID, branchID)
+		return nil, err
+	}, db.WithContext(ctx))
+	return err
+}
+
+func (m *PGRefManager) ListBranches(ctx context.Context, repositoryID RepositoryID, from BranchID) (BranchIterator, error) {
+	return NewBranchIterator(ctx, m.db, repositoryID, IteratorPrefetchSize, string(from)), nil
+}
+
+func (m *PGRefManager) GetCommit(ctx context.Context, repositoryID RepositoryID, commitID CommitID) (*Commit, error) {
+	commit, err := m.db.Transact(func(tx db.Tx) (interface{}, error) {
+		records := make([]*CommitRecord, 0)
+		err := tx.Select(&records, `
+					SELECT id, committer, message, creation_date, parents, tree_id, metadata
+					FROM kv_commits
+					WHERE repository_id = $1 AND id >= $2
+					LIMIT 2`,
+			repositoryID, commitID)
+		if errors.Is(err, db.ErrNotFound) {
+			return nil, ErrNotFound
+		}
+		if err != nil {
+			return nil, err
+		}
+		startWith := make([]*Commit, 0)
+		for _, c := range records {
+			if strings.HasPrefix(string(c.CommitID), string(commitID)) {
+				startWith = append(startWith, c.Commit)
+			}
+		}
+		if len(startWith) == 0 || len(startWith) > 1 {
+			return "", ErrNotFound // empty or ambiguous
+		}
+		return startWith[0], nil
+	}, db.ReadOnly(), db.WithContext(ctx))
+	if errors.Is(err, db.ErrNotFound) {
+		return nil, ErrNotFound
+	}
+	if err != nil {
+		return nil, err
+	}
+	return commit.(*Commit), nil
+}
+
+func (m *PGRefManager) AddCommit(ctx context.Context, repositoryID RepositoryID, commit Commit) (CommitID, error) {
+	_, err := m.db.Transact(func(tx db.Tx) (interface{}, error) {
+		// commits are written based on their content hash, if we insert the same ID again,
+		// it will necessarily have the same attributes as the existing one, so no need to overwrite it
+		_, err := tx.Exec(`
+				INSERT INTO kv_commits 
+				(repository_id, id, committer, message, creation_date, parents, tree_id, metadata)
+				VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
+				ON CONFLICT DO NOTHING`,
```

Comment:
It would be easier to debug weirdness in calls to this layer if we _did_ fail the insert if any values change.

Reply:
open to code suggestions.

Reply:
I feel that making this action idempontent would be better. We should (and do) test elsewhere that a given commit hashes to the same value consistently, and that different commits hash to different values. We'd need to add a lot of code here to
a. figure out that an insert on that PK has already taken place
b. reimplementing the logic of the commit's Identity function (i.e. to compare the stored commit is the same as the one we're attempting to insert).

I don't think the value of easier debugging justifies the added code and complexity